### PR TITLE
fixes for langchain text clients calling the proxy

### PR DESCRIPTION
**Pull request by eslamkarim**
## Issue being addressed
Trying to use langchain text model clients `AzureOpenAI, OpenAI` fails
1. AzureOpenAI uses `/openai/deployments/model_name/completions` to call the models which is not present in the proxy routes
2. OpenAI client fails to respond with a text response and replies with a chat response instead
3. async text completion when talking to azure returns a coroutine instead of a model response (tried to trace what exactly is not being awaited but failed)

## How to reproduce
in order to repro these issues you can use a simple python and langchain script to invoke a message to the litellm proxy

```python
from langchain_openai import AzureOpenAI, OpenAI

azurellm = AzureOpenAI(
    azure_deployment="gpt-35-instruct",
    api_key="sk-eh0S**************2Kry",
    azure_endpoint="http://localhost:4000",
    api_version="2024-02-01"
)
openaillm = OpenAI(
    model="gpt-35-instruct",
    api_key="sk-eh0S5X********************y",
    base_url="http://localhost:4000",
)
print(azurellm.invoke("how are you?"))
print(openaillm.invoke("how are you?"))
```
the above will generate the error messages.
```bash
#from the proxy logs for the azure client
INFO:     127.0.0.1:56904 - "POST /openai/deployments/gpt-35-instruct/completions?api-version=2024-02-01 HTTP/1.1" 404 Not Found
#from proxy logs before switching to normal text_completion for the both clients (tried to track where it's not awaiting the co routine but failed :C )
    raise ValueError(errors) from e
ValueError: [TypeError("'coroutine' object is not iterable"), TypeError('vars() argument must have __dict__ attribute')]

#from proxy logs after switching to normal text_completion for the openai client but not remove kwargs from the call
    raise APIConnectionError(
litellm.exceptions.APIConnectionError: Completions.create() got an unexpected keyword argument 'original_function'

#from proxy logs after switching to normal text_completion for the openai client and remove kwargs from the call but the prompt is being sent as empty string
No logs just gibberish responses

#from langchain client logs without returning a text model response
text
  none is not an allowed value (type=type_error.none.not_allowed)
```

### CODE REVIEW

File: litellm/proxy/proxy_server.py [hunk 1]
```diff
@@ -3064,7 +3073,7 @@ async def completion(
         elif (
             llm_router is not None and data["model"] in router_model_names
         ):  # model in router model list
-            response = await llm_router.atext_completion(**data)
+            response = llm_router.text_completion(**data)
```
Comment (ghost) [inline litellm/proxy/proxy_server.py:3076]: @eslamkarim making an async call sync would really impact perf. 

The text completion issue you were seeing should've been addressed recently. If not, happy to address that in a separate issue.

File: litellm/proxy/proxy_server.py [hunk 2]
```diff
@@ -3106,7 +3115,25 @@ async def completion(
...
+            text_choices = TextChoices()
+            text_choices["text"] = response["choices"][0]["message"]["content"]
+            text_choices["index"] = response["choices"][0]["index"]
+            text_choices["finish_reason"] = response["choices"][0]["finish_reason"]
+            text_response["choices"] = [text_choices]
+            text_response["usage"] = response.get("usage", None)
+            return text_response
+        except Exception as e:
+            verbose_proxy_logger.debug("Error occurred converting to text completion object - Error:  %s",e)
+            return response
```
Comment (ghost) [inline litellm/proxy/proxy_server.py:3136]: this should already be getting handled - https://github.com/BerriAI/litellm/blob/22d56037788ca1ad33055768ad99709925c023c4/litellm/main.py#L3168

**Comment by ghost** [comment]
can repro the format issue - working on it 
Screenshot 2024-03-30 at 8 46 51 AM

**Comment by ghost** [comment]
oh i see - b/c it's async it was missing the translation block. that's a simple fix

**Comment by ghost** [comment]
Fixed. i'll also add validation testing to our ci/cd for preventing any future regressions. 

Screenshot 2024-03-30 at 8 54 21 AM

**Comment by ghost** [comment]
Can confirm this also works with langchain - 
Screenshot 2024-03-30 at 8 55 16 AM

**Comment by eslamkarim** [comment]
@krrishdholakia that's perfect! thank alot krrish you guys are crazy fast :D
can you let me know which version i should be testing? i can retest everything from my end and close this 
even if it's a dev release it should be enough

**Comment by ghost** [comment]
hey @eslamkarim should be live for testing on cli proxy on `1.34.17.dev1` 

Official release running through ci/cd

File: litellm/proxy/proxy_server.py [hunk 3]
```diff
@@ -2991,6 +2992,11 @@ def model_list(
     dependencies=[Depends(user_api_key_auth)],
     tags=["completions"],
 )
+@router.post(
+    "/openai/deployments/{model:path}/completions",
+    dependencies=[Depends(user_api_key_auth)],
+    tags=["completions"],
+)
```
Comment (eslamkarim) [inline litellm/proxy/proxy_server.py:2999]: @krrishdholakia i guess this part is still needed if you use the azure client? it's not present in the dev release
with the query param part as well

**Comment by eslamkarim** [comment]
i also ran a quick test with a vertex model it failed as the prompt doesn't seem to be passed

```bash
llm_model.predict(,**{'tempreature': 0.7, 'top_p': 1, 'max_output_tokens': 256}).text

An error occurred: VertexAIException - 400 The text content is empty.
```

File: litellm/proxy/proxy_server.py (hunk 3 above)
Comment (ghost) [reply litellm/proxy/proxy_server.py:2999]: just added - https://github.com/BerriAI/litellm/commit/0342cd3b6b0aefb1b46e6d1b27248a403d8d7048

**Comment by ghost** [comment]
> i also ran a quick test with a vertex model it failed as the prompt doesn't seem to be passed
> 
> ```shell
> llm_model.predict(,**{'tempreature': 0.7, 'top_p': 1, 'max_output_tokens': 256}).text
> 
> An error occurred: VertexAIException - 400 The text content is empty.
> ```

is that just a random vertex ai error ? there should be no diff b/w calling it on /chat/completions and /completions 

if you run the proxy with `--debug` you should be able to see the exact call being sent to vertex

File: litellm/proxy/proxy_server.py (hunk 3 above)
Comment (eslamkarim) [reply litellm/proxy/proxy_server.py:2999]: what about the api version being passed as a query param?
is it handled somewhere else already?

**Comment by ghost** [comment]
Can confirm prompt is being passed 

## Chat completions
Screenshot 2024-03-30 at 11 34 40 AM
Screenshot 2024-03-30 at 11 34 37 AM


## Completions

Screenshot 2024-03-30 at 11 34 07 AM
Screenshot 2024-03-30 at 11 34 04 AM

**Comment by ghost** [comment]
closing this PR as the issues have now been solved. Let me know if anything remains @eslamkarim

**Comment by eslamkarim** [comment]
will raise a proper issue for vertex
but can we close the api version piece as well?

**Comment by ghost** [comment]
tracking here - https://github.com/BerriAI/litellm/issues/2383

will try and have it done by monday

**Comment by ghost** [comment]
Re: vertex, i don't think it's an issue on our end. Can you run your proxy with --debug and confirm the prompt / params aren't being sent to vertex? It looked like an instability issue on their end.

**Comment by eslamkarim** [comment]
that's fine i will do proper testing and collect more info with vertex models
we should be good on the issues mentioned here thanks a lot for your effort!

**Pull request closed without merging**